File: src/main/java/org/funivan/lologa/algo/ga/genome/value/LockedTilesValue.java
package org.funivan.lologa.algo.ga.genome.value;

import org.funivan.lologa.algo.find.HandlerInterface;
import org.funivan.lologa.algo.find.multiple.LockedTilesFinder;
import org.funivan.lologa.algo.gameplay.GameplayInterface;
import org.funivan.lologa.tiles.TilesInterface;

public class LockedTilesValue implements ValueInterface {

    private final HandlerInterface finder;

    public LockedTilesValue(GameplayInterface gameplay) {
        this.finder = new LockedTilesFinder(gameplay.minimum());
    }

    @Override
    public String type() {
        return "locked";
    }

    @Override
    public Double value(TilesInterface original, TilesInterface tiles) {
        final double oldLocked = (double) this.finder.handle(original).size() / (double) original.size();
        final double newLocked = (double) this.finder.handle(tiles).size() / (double) tiles.size();
        return oldLocked - newLocked;
    }

}

File: src/main/java/org/funivan/lologa/algo/ga/genome/value/LockedTilesValue.kt
package org.funivan.lologa.algo.ga.genome.value

import org.funivan.lologa.algo.find.TilesAction
import org.funivan.lologa.algo.find.multiple.LockedTilesFinder
import org.funivan.lologa.algo.gameplay.GameplayInterface
import org.funivan.lologa.tiles.TilesInterface

class LockedTilesValue(gameplay: GameplayInterface) : ValueInterface {

    private val finder: TilesAction

    init {
        this.finder = LockedTilesFinder(gameplay.minimum())
    }

    override fun type(): String {
        return "locked"
    }

    override fun value(original: TilesInterface, tiles: TilesInterface): Double {
        val oldLocked = this.finder.perform(original).size().toDouble() / original.size().toDouble()
        val newLocked = this.finder.perform(tiles).size().toDouble() / tiles.size().toDouble()
        return oldLocked - newLocked
    }

}
